### PR TITLE
[IMP]stock: add user and delivery contact as follower

### PR DESCRIPTION
**Pull request by yhu-odoo**
=======
PURPOSE
=======

Avoid communication being lost on transfers due to the fact that nobody is following.

=============
SPECIFICATION
=============

When setting a delivery address, add the contact in question as a follower of the transfer.
When a user validates a transfer, he should be automatically added as a follower of this transfer.

===========
LINKS
===========

task 2172675
pr #44350
--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by amoyaux** [comment]
To discuss with @oma-odoo 
When the warehouse is set as 3 steps delivery then the partner is added in the 3 pickings.
However he don't really care about pack or output zone. What really matters is the delivery to him so I think it should just be added in the delivery picking to customer location and not the 3 pickings.

**Comment by oma-odoo** [comment]
@amoyaux Yes it would be better if the customer was only added on the Ship, can it be done easily?

File: addons/stock/models/stock_picking.py [hunk 1]
```diff
@@ -570,11 +570,23 @@ def create(self, vals):
...
+                picking.message_subscribe([vals.get('partner_id')])
+
         return res
 
     def write(self, vals):
         if vals.get('picking_type_id') and self.state != 'draft':
             raise UserError(_("Changing the operation type of this record is forbidden at this point."))
+        # set partner as a follower and unfollow old partner
+        if vals.get('partner_id'):
+            for picking in self:
```
Comment (amoyaux) [inline addons/stock/models/stock_picking.py]:
``` python3
if not picking.location_dest_id.usage == 'customer':
    continue
```

File: addons/stock/models/stock_picking.py [hunk 2]
```diff
@@ -570,11 +570,23 @@ def create(self, vals):
                         move[2]['company_id'] = picking_type.company_id.id
         res = super(Picking, self).create(vals)
         res._autoconfirm_picking()
+
+        # set partner as follower
+        if vals.get('partner_id'):
+            for picking in res:
```
Comment (amoyaux) [inline addons/stock/models/stock_picking.py]:
```python3
res.filtered(lambda p: p.location_dest_id.usage == 'customer')
```

File: addons/stock/models/stock_picking.py (hunk 1 above)
Comment (yhu-odoo) [reply addons/stock/models/stock_picking.py]: Oma said partner should be add as a follower for both delivery and receipt, so I think we should check `picking.location_id.usage == 'supplier' or picking.location_dest_id.usage == 'customer' `. right?

Comment (amoyaux) [reply addons/stock/models/stock_picking.py]: @yhu-odoo :+1:

**Comment by sle-odoo** [comment]
robodoo r+

**Pull request closed without merging**